sys/arm64/vmm/vmm_dev.c

```diff
@@ -0,0 +1,1041 @@
+	if (error)
+		return (error);
+
+	sc = vmmdev_lookup2(cdev);
+	if (sc == NULL)
+		return (ENXIO);
+
+	error = 0;
+	vcpuid = -1;
+	vcpu = NULL;
+	vcpus_locked = NONE;
+	memsegs_locked = false;
+
+	/*
+	 * Some VMM ioctls can operate only on vcpus that are not running.
+	 */
+	switch (cmd) {
+	case VM_RUN:
+	case VM_GET_REGISTER:
+	case VM_SET_REGISTER:
+	case VM_GET_REGISTER_SET:
+	case VM_SET_REGISTER_SET:
+	case VM_INJECT_EXCEPTION:
+	case VM_GET_CAPABILITY:
+	case VM_SET_CAPABILITY:
+	case VM_GLA2GPA_NOFAULT:
+	case VM_ACTIVATE_CPU:
+		/*
+		 * ioctls that can operate only on vcpus that are not running.
+		 */
+		vcpuid = *(int *)data;
+		vcpu = vm_alloc_vcpu(sc->vm, vcpuid);
+		if (vcpu == NULL) {
+			error = EINVAL;
+			goto done;
+		}
+		error = vcpu_lock_one(vcpu);
+		if (error)
+			goto done;
+		vcpus_locked = SINGLE;
+		break;
+
+	case VM_ALLOC_MEMSEG:
+	case VM_MMAP_MEMSEG:
+	/* TODO: VM_MUNMAP_MEMSEG */
+	case VM_REINIT:
+	case VM_ATTACH_VGIC:
+		/*
+		 * ioctls that modify the memory map must lock memory
+		 * segments exclusively.
+		 */
+		vm_xlock_memsegs(sc->vm);
+		memsegs_locked = true;
+		/* FALLTHROUGH */
+	/* TODO: Any non-memseg ioctls? */
+		/*
+		 * ioctls that operate on the entire virtual machine must
+		 * prevent all vcpus from running.
+		 */
+		error = vcpu_lock_all(sc);
+		if (error)
+			goto done;
+		vcpus_locked = ALL;
+		break;
+	case VM_GET_MEMSEG:
+	case VM_MMAP_GETNEXT:
+		/*
+		 * Lock the memory map while it is being inspected.
+		 */
+		vm_slock_memsegs(sc->vm);
+		memsegs_locked = true;
+		break;
+
+	case VM_STATS:
+		/*
+		 * These do not need the vCPU locked but do operate on
+		 * a specific vCPU.
+		 */
+		vcpuid = *(int *)data;
+		vcpu = vm_alloc_vcpu(sc->vm, vcpuid);
+		if (vcpu == NULL) {
+			error = EINVAL;
+			goto done;
+		}
+		break;
+
+	case VM_SUSPEND_CPU:
+	case VM_RESUME_CPU:
+		/*
+		 * These can either operate on all CPUs via a vcpuid of
+		 * -1 or on a specific vCPU.
+		 */
+		vcpuid = *(int *)data;
+		if (vcpuid == -1)
+			break;
+		vcpu = vm_alloc_vcpu(sc->vm, vcpuid);
+		if (vcpu == NULL) {
+			error = EINVAL;
+			goto done;
+		}
+		break;
+
+	case VM_ASSERT_IRQ:
+		vi = (struct vm_irq *)data;
+		error = vm_assert_irq(sc->vm, vi->irq);
+		break;
+	case VM_DEASSERT_IRQ:
+		vi = (struct vm_irq *)data;
+		error = vm_deassert_irq(sc->vm, vi->irq);
+		break;
+	default:
+		break;
+	}
+
+	switch (cmd) {
+	case VM_RUN: {
+		struct vm_exit *vme;
+
+		vmrun = (struct vm_run *)data;
+		vme = vm_exitinfo(vcpu);
+
+		error = vm_run(vcpu);
+		if (error != 0)
+			break;
+
+		error = copyout(vme, vmrun->vm_exit, sizeof(*vme));
+		if (error != 0)
+			break;
+		break;
+	}
+	case VM_SUSPEND:
+		vmsuspend = (struct vm_suspend *)data;
+		error = vm_suspend(sc->vm, vmsuspend->how);
+		break;
+	case VM_REINIT:
+		error = vm_reinit(sc->vm);
+		break;
+	case VM_STAT_DESC: {
+		statdesc = (struct vm_stat_desc *)data;
+		error = vmm_stat_desc_copy(statdesc->index,
+					statdesc->desc, sizeof(statdesc->desc));
```

Comment:
There are a bunch of small indentation style issues throughout this switch statement.

Reply:
This is also from amd64. IMO it's not worth trying to fix here, as that'll make it harder to reconcile the two versions.